### ISSUE TITLE
"Bad file descriptor" error

### ISSUE THREAD

**Issue by schrauf** [kdheepak/Termios.jl#5]
Hi, I saw your (great) presentation today and wanted to try TerminalUserInterfaces.jl
but I'm having a "Bad file descriptor" error. posted the issue here since I think my problem starts with TERMIOS.
I don't understand it, sorry if it is obvious.

Working in linux (linux mint 19.3 64 bit).
I copied the usage example from the docs into a file "termios-debug.jl":

```julia
using TERMIOS
const T = TERMIOS
term_stdin = T.termios()
T.tcgetattr(stdin, term_stdin)
term_stdin.c_iflag |= T.IGNBRK
T.tcsetattr(stdin, T.TCSANOW, term_stdin)
```

```
user@host:dir$ julia termios-debug.jl 
ERROR: LoadError: TERMIOSError: tcgetattr failed: Bad file descriptor
Stacktrace:
 [1] tcgetattr(::RawFD, ::TERMIOS.termios) at ~/.julia/packages/TERMIOS/Kvyl9/src/TERMIOS.jl:610
 [2] tcgetattr(::Base.TTY, ::TERMIOS.termios) at ~/.julia/packages/TERMIOS/Kvyl9/src/TERMIOS.jl:612
 [3] top-level scope at dir/termios-debug.jl:4
 [4] include(::Module, ::String) at ./Base.jl:377
 [5] exec_options(::Base.JLOptions) at ./client.jl:288
 [6] _start() at ./client.jl:484
in expression starting at dir/termios-debug.jl:4
```

**Comment by kdheepak** [comment]
I've experienced that error on my Mac as well. For me it is intermittent/inconsistent. Some runs, everything works fine. Are you experiencing this error every time you run it?

I've also been able to decrease the number of times I get this by moving the code that calls `tcgetattr` into a function. Can you try this and see if it makes a difference?

```julia
using TERMIOS
const T = TERMIOS

function main()
    term_stdin = T.termios()
    T.tcgetattr(stdin, term_stdin)
    term_stdin.c_iflag |= T.IGNBRK
    T.tcsetattr(stdin, T.TCSANOW, term_stdin)
end

main()
```

Or just in a `repl`?

Your code was erroring in the first call to `tcgettr`. And the bad file descriptor means that it is not able to `tcgetattr` on the `stdin` file, and I'm not sure why that would be the case.

**Comment by schrauf** [comment]
Thanks for your time and the fast response!
I'm experiencing this error every time (which might make it easier to debug :)

I've just tried your suggestion both as a file and in the repl, and I get the same error.
In case it helps, my stdin works:

```
julia> print(stdin, "here")
here
julia> stdin
Base.TTY(RawFD(0x0000000a) paused, 0 bytes waiting)
```

**Comment by kdheepak** [comment]
Interesting. Can you try the following in python?

```python
import termios, sys
fd = sys.stdin.fileno()
print(fd)
termios.tcgetattr(fd)
```

**Comment by schrauf** [comment]
Yes, 

```python
In [1]: import termios, sys 
   ...: fd = sys.stdin.fileno() 
   ...: print(fd) 
   ...: termios.tcgetattr(fd)                                                                                                          
0
Out[1]: 
[16640,
 5,
 191,
 35387,
 15,
 15,
 [b'\x03',
  b'\x1c',
  b'\x7f',
...
b'\x00']]
```

the last element of a list is itself a list with 32 byte elements

**Comment by kdheepak** [comment]
In Python, this is the source:

https://github.com/python/cpython/blob/ba18c0b13ba3c08077ea3db6658328523823a33f/Modules/termios.c#L81-L96

And in julia, this is the source:

https://github.com/kdheepak/TERMIOS.jl/blob/37d991d53cfe4bd337656fc6633f11c8b2b429b2/src/TERMIOS.jl#L601-L613

You can see they are pretty much identical.

Now, the only thing that could be different is the layout of the `termios` struct itself.

Can you print `termios.tcgetattr(fd)` in python, and print `T.termios()` in julia? 

If there's a difference in the layout, then it is something that can be fixed in this package. If there's no difference in the layout, we'll have to ping someone on the core Julia team who has knowledge of libuv and Julia's IO system for more information about what is going on.

**Comment by kdheepak** [comment]
Interestingly the tests on Travis are passing:

https://travis-ci.com/github/kdheepak/TERMIOS.jl/jobs/287020194#L153

https://github.com/kdheepak/TERMIOS.jl/blob/37d991d53cfe4bd337656fc6633f11c8b2b429b2/test/runtests.jl#L56-L64

**Comment by c42f** [comment]
Regarding layout, the result of using `gcc -E` with the following test.c:

```C
#include <termios.h>
#include <unistd.h>
```

indicates that on my system we have

```C
typedef unsigned char cc_t;
typedef unsigned int speed_t;
typedef unsigned int tcflag_t;

struct termios
  {
    tcflag_t c_iflag;
    tcflag_t c_oflag;
    tcflag_t c_cflag;
    tcflag_t c_lflag;
    cc_t c_line;
    cc_t c_cc[32];
    speed_t c_ispeed;
    speed_t c_ospeed;
  };
```

All looks ABI compatible when compared to `TERMIOS.termios`

**Comment by c42f** [comment]
Is this a problem?

```julia
julia> TERMIOS._file_handle(stdout)
RawFD(0x61636d20)

julia> stdout
Base.TTY(RawFD(0x0000000d) open, 0 bytes waiting)
```

The `RawFD`'s don't match here. There's `Base._fd` which can be used to return `RawFD(0x0000000d)`, but replacing `_file_handle` with that leads the system to complain about `ioctl`s instead. Which might be progress, or maybe not.

**Comment by schrauf** [comment]
Ok, I also have a mismatch in my personal computer.

```julia
julia> using TERMIOS

julia> TERMIOS._file_handle(stdout)
RawFD(0x20657370)

julia> stdout
Base.TTY(RawFD(0x0000000d) open, 0 bytes waiting)
```

In a remote server I also have a mismatch, but with a different output for `_file_handle(stdout)`.

```julia
julia> using TERMIOS

julia> TERMIOS._file_handle(stdout)
RawFD(0x00000000)

julia> stdout
Base.TTY(RawFD(0x0000000d) open, 0 bytes waiting)
```

If I try in the server the code suggested by @kdheepak [here](https://github.com/kdheepak/TERMIOS.jl/issues/5#issuecomment-665932805), then I don't get the "Bad file descriptor" error. But later when trying examples from TerminalUserInterface.jl I get the complain about `ioctl`s which @c42f mentions.

**Comment by kdheepak** [comment]
> ```
> julia> TERMIOS._file_handle(stdout)
> RawFD(0x61636d20)
>
> julia> stdout
> Base.TTY(RawFD(0x0000000d) open, 0 bytes waiting)
> ```

So this is definitely the issue, right? 

https://github.com/kdheepak/TERMIOS.jl/blob/37d991d53cfe4bd337656fc6633f11c8b2b429b2/src/TERMIOS.jl#L599

Any idea what we should use instead?

**Comment by c42f** [comment]
Well, there's `Base._fd` as I noted above. It's possible this is correct, though i didn't look into it in detail.

**Comment by kdheepak** [comment]
Is `IOCTL` error related to this issue?

https://github.com/kdheepak/TerminalUserInterfaces.jl/issues/3

If so, I can make the `Base._fd` changes and make a new release. It still will need fixing the `IOCTL`, but at least it is progress. I'll also set up a Linux VM so that I can debug this locally.

**Comment by kdheepak** [comment]
Surprisingly I'm unable to reproduce this issue on my Ubuntu VM.

![Screen Shot 2020-09-06 at 4 57 22 PM](https://user-images.githubusercontent.com/1813121/92337077-2cf6f780-f064-11ea-956a-2aa29143e5e9.png)

But I've gone ahead updated to use `Base._fd`. I tested that on my Mac and on Ubuntu, and it seems like it works. But was working for me before I made the change as well.

I just submitted a registration request for a new version (v0.2.1) of this package: https://github.com/kdheepak/TERMIOS.jl/commit/3c32ffde07fcc9223e7fe139afb7040d460074f6. When that is merged and tagged, if others can try it out that would be great.
